# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: media/webrtc/signaling/src/mediapipeline/MediaPipeline.cpp
# Commit: 0b03dd9d20ac
# Full Hash: 0b03dd9d20ace235502714101097fe34d1864563
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2019-07-31 15:46:43
# Regressor Bug: 1493613
# File Overlap Count: 1
# Description:
#   Bug 1493613 - Update muted state through MediaStreamTrackSource. r=bwc,smaug
#   
#   This ensures all clones of the original track also receives the new muted state.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D37933
# ==============================================================================

diff -r cf84eeffed22 -r 0b03dd9d20ac media/webrtc/signaling/src/mediapipeline/MediaPipeline.cpp
--- a/media/webrtc/signaling/src/mediapipeline/MediaPipeline.cpp	Tue Jul 30 14:55:48 2019 +0000
+++ b/media/webrtc/signaling/src/mediapipeline/MediaPipeline.cpp	Wed Jul 31 09:35:06 2019 +0000
@@ -23,6 +23,7 @@
 #include "MediaStreamGraphImpl.h"
 #include "MediaStreamListener.h"
 #include "MediaStreamTrack.h"
+#include "RemoteTrackSource.h"
 #include "RtpLogger.h"
 #include "VideoFrameConverter.h"
 #include "VideoSegment.h"
@@ -1164,7 +1165,8 @@
 class GenericReceiveListener : public MediaStreamTrackListener {
  public:
   explicit GenericReceiveListener(dom::MediaStreamTrack* aTrack)
-      : mTrack(aTrack),
+      : mTrack(new nsMainThreadPtrHolder<dom::MediaStreamTrack>(
+            "GenericReceiveListener::mTrack", aTrack)),
         mTrackId(aTrack->GetInputTrackId()),
         mSource(mTrack->GetInputStream()->AsSourceStream()),
         mPrincipalHandle(PRINCIPAL_HANDLE_NONE),
@@ -1173,10 +1175,7 @@
     MOZ_RELEASE_ASSERT(mSource, "Must be used with a SourceMediaStream");
   }
 
-  virtual ~GenericReceiveListener() {
-    NS_ReleaseOnMainThreadSystemGroup("GenericReceiveListener::track_",
-                                      mTrack.forget());
-  }
+  virtual ~GenericReceiveListener() = default;
 
   void AddTrackToSource(uint32_t aRate = 0) {
     MOZ_ASSERT((aRate != 0 && mTrack->AsAudioStreamTrack()) ||
@@ -1226,8 +1225,8 @@
   }
 
   void OnRtpReceived_m() {
-    if (mListening && mTrack->Muted()) {
-      mTrack->MutedChanged(false);
+    if (mListening) {
+      static_cast<RemoteTrackSource&>(mTrack->GetSource()).SetMuted(false);
     }
   }
 
@@ -1268,7 +1267,7 @@
   }
 
  protected:
-  RefPtr<dom::MediaStreamTrack> mTrack;
+  const nsMainThreadPtrHandle<dom::MediaStreamTrack> mTrack;
   const TrackID mTrackId;
   const RefPtr<SourceMediaStream> mSource;
   PrincipalHandle mPrincipalHandle;